pilot/cmd/sidecar-injector/main.go

```diff
@@ -131,25 +132,59 @@ func patchCertLoop() error {
 	if err != nil {
 		return err
 	}
-
 	watchDir, _ := filepath.Split(flags.caCertFile)
 	if err = watcher.Watch(watchDir); err != nil {
 		return fmt.Errorf("could not watch %v: %v", flags.caCertFile, err)
 	}
 
+	if err = util.PatchMutatingWebhookConfig(client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations(),
+		flags.webhookConfigName, flags.webhookName, caCertPem); err != nil {
+		return err
+	}
+
+	shouldPatch := make(chan struct{})
+
+	watchlist := cache.NewListWatchFromClient(
+		client.AdmissionregistrationV1beta1().RESTClient(),
+		"mutatingwebhookconfigurations",
+		"",
+		fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", flags.webhookConfigName)))
+
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1beta1.MutatingWebhookConfiguration{},
+		time.Second*0,
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				config := newObj.(*v1beta1.MutatingWebhookConfiguration)
+				for i, w := range config.Webhooks {
+					if w.Name == flags.webhookName {
+						if !bytes.Equal(config.Webhooks[i].ClientConfig.CABundle, caCertPem) {
+							log.Infof("Detected a change in CABundle, patching MutatingWebhookConfiguration again")
+							shouldPatch <- struct{}{}
+							break
+						}
+					}
+				}
+			},
+		},
+	)
+	go controller.Run(stopCh)
+
 	go func() {
-		tickerC := time.NewTicker(time.Second).C
 		for {
 			select {
-			case <-tickerC:
+			case <-shouldPatch:
 				if err = util.PatchMutatingWebhookConfig(client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations(),
 					flags.webhookConfigName, flags.webhookName, caCertPem); err != nil {
 					log.Errorf("Patch webhook failed: %v", err)
 				}
 
 			case <-watcher.Event:
 				if b, err := ioutil.ReadFile(flags.caCertFile); err == nil {
+					log.Infof("Detected a change in CABundle (via secret), patching MutatingWebhookConfiguration again")
 					caCertPem = b
+					shouldPatch <- struct{}{}
```

Comment:
Is there a need to do it this way rather than just to the patch here?

Reply:
no strong reason, just to avoid code duplication

Reply:
That can be solved with a helper function - triggering through a channel is unnecessarily complex here, unless I'm missing something.